packages/gatsby-script/src/gatsby-script.tsx

```diff
@@ -0,0 +1,181 @@
+import React, { useEffect, useContext } from "react"
+import { PartytownContext } from "./partytown-context"
+import type { ReactElement, ScriptHTMLAttributes } from "react"
+import { requestIdleCallback } from "./request-idle-callback-shim"
+
+export enum ScriptStrategy {
+  postHydrate = `post-hydrate`,
+  idle = `idle`,
+  offMainThread = `off-main-thread`,
+}
+
+// eslint-disable-next-line @typescript-eslint/naming-convention
+export interface ScriptProps
+  extends Omit<ScriptHTMLAttributes<HTMLScriptElement>, `onLoad` | `onError`> {
+  id?: string
+  strategy?: ScriptStrategy | `post-hydrate` | `idle` | `off-main-thread`
+  children?: string
+  onLoad?: (event: Event) => void
+  onError?: (event: ErrorEvent) => void
+}
+
+const handledProps = new Set([
+  `src`,
+  `strategy`,
+  `dangerouslySetInnerHTML`,
+  `children`,
+  `onLoad`,
+  `onError`,
+])
+
+export const scriptCache = new Set()
+
+export function Script(props: ScriptProps): ReactElement | null {
+  const {
+    src,
+    strategy = ScriptStrategy.postHydrate,
+    onLoad,
+    onError,
+  } = props || {}
+  const { collectScript } = useContext(PartytownContext)
+
+  useEffect(() => {
+    let script: HTMLScriptElement | null
+
+    switch (strategy) {
+      case ScriptStrategy.postHydrate:
+        script = injectScript(props)
+        break
+      case ScriptStrategy.idle:
+        requestIdleCallback(() => {
+          script = injectScript(props)
+        })
+        break
+      case ScriptStrategy.offMainThread:
+        if (typeof window !== `undefined` && collectScript) {
+          const attributes = resolveAttributes(props)
+          collectScript(attributes)
+        }
+        break
+    }
+
+    return (): void => {
+      if (onLoad) {
+        script?.removeEventListener(`load`, onLoad)
+      }
+      if (onError) {
+        script?.removeEventListener(`error`, onError)
+      }
+      script?.remove()
+    }
+  }, [])
+
+  if (strategy === ScriptStrategy.offMainThread) {
+    const inlineScript = resolveInlineScript(props)
+    const attributes = resolveAttributes(props)
+
+    if (typeof window === `undefined` && collectScript) {
+      collectScript(attributes)
+    }
```

Comment:
We should show warning if `collectScript` is not available (which most likely mean user is setting it in unsupported location / below our context provider / outside of main react tree)